processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,375 @@
+		}
+	}
+	return key, newLabelValues
+}
+
+// aggregateTimeseriesGroups attempts to merge each group of timesreis in keyToTimeseriesMap through the spefic calculation indicated by aggrType and dataType to construct the new timeseries with new label values indicated in keyToLabelValuesMap
+// Returns the new list of timeseries
+func (mtp *metricsTransformProcessor) aggregateTimeseriesGroups(keyToTimeseriesMap map[string][]*metricspb.TimeSeries, keyToLabelValuesMap map[string][]*metricspb.LabelValue, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.TimeSeries {
+	newTimeSeriesList := make([]*metricspb.TimeSeries, len(keyToTimeseriesMap))
+	idxCounter := 0
+	for key, element := range keyToTimeseriesMap {
+		timestampToPoints, startTimestamp := mtp.analyzeTimeseriesForAggregation(element)
+		newPoints := mtp.aggregatePoints(timestampToPoints, aggrType, dataType)
+		// newSingleTimeSeries is an aggregated timeseries
+		newSingleTimeSeries := &metricspb.TimeSeries{
+			StartTimestamp: startTimestamp,
+			LabelValues:    keyToLabelValuesMap[key],
+			Points:         newPoints,
+		}
+		newTimeSeriesList[idxCounter] = newSingleTimeSeries
+		idxCounter++
+	}
+
+	return newTimeSeriesList
+}
+
+// analyzeTimeseriesForAggregation uses timestamp as an indicator to group aggregatable points in the group of timeseries, and determines the final value of the start timestamp for this groups of timeseries.
+// Returns a map that groups points by timestamps and the final start timestamp
+func (mtp *metricsTransformProcessor) analyzeTimeseriesForAggregation(timeseries []*metricspb.TimeSeries) (map[string][]*metricspb.Point, *timestamp.Timestamp) {
+	var startTimestamp *timestamp.Timestamp
+	// timestampToPoints maps from timestamp string to aggregatable points
+	timestampToPoints := make(map[string][]*metricspb.Point)
+	for _, ts := range timeseries {
+		if startTimestamp == nil || ts.StartTimestamp.Seconds < startTimestamp.Seconds || (ts.StartTimestamp.Seconds == startTimestamp.Seconds && ts.StartTimestamp.Nanos < startTimestamp.Nanos) {
+			startTimestamp = ts.StartTimestamp
+		}
+		for _, p := range ts.Points {
+			if points, ok := timestampToPoints[p.Timestamp.String()]; ok {
+				timestampToPoints[p.Timestamp.String()] = append(points, p)
+			} else {
+				timestampToPoints[p.Timestamp.String()] = []*metricspb.Point{p}
+			}
+		}
+	}
+	return timestampToPoints, startTimestamp
+}
+
+// aggregatePoints aggregates points in the groups provided in timestampToPoints by the specific caluculation indicated by aggrType and dataType
+// Returns a group of aggregated points
+func (mtp *metricsTransformProcessor) aggregatePoints(timestampToPoints map[string][]*metricspb.Point, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.Point {
+	// initialize to size 0. unsure how many points will come out because distribution points are not guaranteed to be aggregatable. (mismatched bounds)
```

Comment:
You know the array will be at least `len(timestampToPoints)` so you can initialize the capacity of the slice to this value. I also think is more reason to split this into three separate functions

Reply:
Wiil do! Thanks!